src/CistromeToolkit.js

```diff
@@ -32,197 +45,487 @@ export default function CistromeToolkit(props) {
+    }, [selectedRequestIndex]);
 
+    // Check if parameters for each API is ready and can make a good URL
     useEffect(() => {
-        let didUnmount = false;
-        if(intervalParams) {
-            setRequestStatus({ msg: "Receiving Cistrome DB API response...", isLoading: true });
+        setIsLatestIntervalParamsReady(validateIntervalParams(latestIntervalParams).success);
+    }, [latestIntervalParams]);
 
-            const {
-                assembly, 
-                chrStartName, 
-                chrStartPos, 
-                chrEndName, 
-                chrEndPos
-            } = intervalParams;
-
-            requestIntervalTFs(assembly, chrStartName, chrStartPos, chrEndName, chrEndPos)
-                .then(([rows, columns]) => {
-                    if(didUnmount) return;
-
-                    // TODO: Consider supporting multiple API queries.
-                    const customColumnMap = {
-                        GSM: "GEO/ENCODE ID",
-                        DCid: "CistromeDB ID",
-                        factor: "Factor",
-                        cellLine: "Cell Line",
-                        CellType: "Cell Type",
-                        species: "Species",
-                        OverlapRatio: "Overlap Ratio",
-                        OverlapPeakNumber: "Overlap Peak Number",
-                    }
-                    const customRows = rows.map(r => {
-                        const newRow = {};
-                        Object.keys(customColumnMap).forEach(k => {
-                            if(customColumnMap[k]) {
-                                newRow[customColumnMap[k]] = r[k];
-                            }
-                        });
-                        return newRow;
-                    });
-                    const costomColumns = Object.values(customColumnMap);
+    useEffect(() => {
+        setIsLatestGeneParamsReady(validateGeneParams(latestGeneParams).success);
+    }, [latestGeneParams]);
 
-                    const msg = `For interval ${chrStartName}:${chrStartPos}-${chrEndPos}`;
-                    setRequestStatus({ msg, isLoading: false });
+    useEffect(() => {
+        setIsLatestPeaksetParamsReady(validatePeaksetParams(latestPeaksetParams).success);
+    }, [latestPeaksetParams]);
+
+    function addRequestHistory(api, parameter, columns, rows) {
+        const isIdentical = {
+            [CISTROME_API_TYPES.INTERVAL]: (d1, d2) => (
+                d1.assembly === d2.assembly 
+                && d1.chrStartName === d2.chrStartName 
+                && d1.chrStartPos === d2.chrStartPos
+                && d1.chrEndName === d2.chrEndName 
+                && d1.chrEndPos === d2.chrEndPos
+            ),
+            [CISTROME_API_TYPES.GENE]: (d1, d2) => (
+                d1.assembly === d2.assembly
+                & d1.distance === d2.distance
+                & d1.gene === d2.gene
+            )
+            // TODO: Support Peak Set API
+        }
```

Comment:
I wonder if this could be replaced with `isEqual(d1, d2)` using the `lodash/isEqual`? https://lodash.com/docs/4.17.15#isEqual